src/runtime/driver/handle.rs

```diff
@@ -0,0 +1,286 @@
+//! The weak handle should be used by anything which is stored in the driver or does not need to
+//! keep the driver alive for it's duration.
+
+use io_uring::{cqueue, squeue};
+use std::cell::RefCell;
+use std::io;
+use std::os::unix::io::{AsRawFd, RawFd};
+use std::rc::{Rc, Weak};
+use std::task::{Context, Poll};
+
+use crate::buf::fixed::FixedBuffers;
+use crate::runtime::driver::op::{Completable, Lifecycle, MultiCQEFuture, Op, Updateable};
+use crate::runtime::driver::Driver;
+
+#[derive(Clone)]
+pub(crate) struct Handle {
+    pub(super) inner: Rc<RefCell<Driver>>,
+}
+
+#[derive(Clone)]
+pub(crate) struct WeakHandle {
+    inner: Weak<RefCell<Driver>>,
+}
+
+impl Handle {
+    pub(crate) fn new(b: &crate::Builder) -> io::Result<Self> {
+        Ok(Self {
+            inner: Rc::new(RefCell::new(Driver::new(b)?)),
+        })
+    }
+
+    pub(crate) fn tick(&self) {
+        self.inner.borrow_mut().tick()
+    }
+
+    pub(crate) fn flush(&self) -> io::Result<usize> {
+        self.inner.borrow_mut().uring.submit()
+    }
+
+    pub(crate) fn register_buffers(
+        &self,
+        buffers: Rc<RefCell<dyn FixedBuffers>>,
+    ) -> io::Result<()> {
+        let mut driver = self.inner.borrow_mut();
+
+        driver
+            .uring
+            .submitter()
+            .register_buffers(buffers.borrow().iovecs())?;
+
+        driver.fixed_buffers = Some(buffers);
+        Ok(())
+    }
+
+    pub(crate) fn unregister_buffers(
+        &self,
+        buffers: Rc<RefCell<dyn FixedBuffers>>,
+    ) -> io::Result<()> {
+        let mut driver = self.inner.borrow_mut();
+
+        if let Some(currently_registered) = &driver.fixed_buffers {
+            if Rc::ptr_eq(&buffers, currently_registered) {
+                driver.uring.submitter().unregister_buffers()?;
+                driver.fixed_buffers = None;
+                return Ok(());
+            }
+        }
+        Err(io::Error::new(
+            io::ErrorKind::Other,
+            "fixed buffers are not currently registered",
+        ))
+    }
+
+    /// Submit an operation to uring.
+    ///
+    /// `state` is stored during the operation tracking any state submitted to
+    /// the kernel.
+    pub(crate) fn submit_op<T, S, F>(&self, mut data: T, f: F) -> io::Result<Op<T, S>>
+    where
+        T: Completable,
+        F: FnOnce(&mut T) -> squeue::Entry,
+    {
+        let mut driver = self.inner.borrow_mut();
+        let index = driver.ops.insert();
+
+        // Configure the SQE
+        let sqe = f(&mut data).user_data(index as _);
+
+        // Create the operation
+        let op = Op::new(self.into(), data, index);
+
+        // Push the new operation
+        while unsafe { driver.uring.submission().push(&sqe).is_err() } {
+            // If the submission queue is full, flush it to the kernel
+            driver.submit()?;
+        }
+
+        Ok(op)
+    }
+
+    pub(crate) fn poll_op<T>(&self, op: &mut Op<T>, cx: &mut Context<'_>) -> Poll<T::Output>
```

Comment:
I don't see the benefit of moving the two big Future functionalities into the new handle.rs file but maybe I just haven't gotten used to it yet. The idea of a Handle is going to get muddy if all the little details have to be put into this file though (it seems to me). Please take all this with a grain of salt, I haven't tried to get it to compile any other way.

Reply:
I'm trying to keep the logic around the driver self-contained to the driver and handle. I want the Op to not have to know about the guts of the driver, just how to communicate with APIs provided by the driver.